File: src/main/resources/static/admin/js/systemInfo.js
$(function(){
    var websocket = null;
    //开启连接
    var url = "ws://" + window.location.host  + "/systemInfo";
    //判断当前浏览器是否支持WebSocket
    if(websocket == null){
        if('WebSocket' in window){
            websocket = new WebSocket(url);
        }else if('MozWebSocket' in window){
            websocket = new MozWebSocket(url);
        }else{
            websocket = new SockJS(url);
        }
    }
    //连接发生错误的回调方法
    websocket.onerror = function(){}
    //连接成功建立的回调方法
    websocket.onopen = function(event){}
    //接收到消息的回调方法
    websocket.onmessage = function(event){
        var json = $.parseJSON(event.data);
        // 服务器信息
        var serverinfo = json.serverinfo;
        if(serverinfo){
            $("#system-info-1").html("<p class=\"text-left\">本地ip地址："+serverinfo.hostIp+"</p>" +
                "<p class=\"text-left\">本地主机名："+serverinfo.hostName+"</p>" +
                "<p class=\"text-left\">操作系统的名称："+serverinfo.osName+"</p>" +
                "<p class=\"text-left\">操作系统的构架："+serverinfo.arch+"</p>" +
                "<p class=\"text-left\">操作系统的版本："+serverinfo.osVersion+"</p>" +
                "<p class=\"text-left\">JVM可以使用的处理器个数："+serverinfo.processors+"</p>" +
                "<p class=\"text-left\">Java的运行环境版本："+serverinfo.javaVersion+"</p>" +
                "<p class=\"text-left\">Java的运行环境供应商："+serverinfo.vendor+"</p>" +
                "<p class=\"text-left\">Java供应商的URL："+serverinfo.javaUrl+"</p>" +
                "<p class=\"text-left\">Java的安装路径："+serverinfo.javaHome+"</p>" +
                "<p class=\"text-left\">默认的临时文件路径："+serverinfo.tmpdir+"</p>");
        }
        // 内存信息
        var memory = json.memory;
        if(memory){
            $("#system-info-2").html("<p class=\"text-left\">java总内存："+memory.jvmTotal+"</p>" +
                "<p class=\"text-left\">JVM剩余内存："+memory.jvmFree+"</p>" +
                "<p class=\"text-left\">JVM使用内存："+memory.jvmUse+"</p>" +
                "<p class=\"text-left\">JVM使用率："+memory.jvmUsage+"</p>" +
                "<p class=\"text-left\">内存总量："+memory.ramTotal+"</p>" +
                "<p class=\"text-left\">当前内存使用量："+memory.ramUse+"</p>" +
                "<p class=\"text-left\">当前内存剩余量："+memory.ramFree+"</p>" +
                "<p class=\"text-left\">内存使用率："+memory.ramUsage+"</p>" +
                "<p class=\"text-left\">交换区总量："+memory.swapTotal+"</p>" +
                "<p class=\"text-left\">当前交换区使用量："+memory.swapUse+"</p>" +
                "<p class=\"text-left\">当前交换区剩余量："+memory.swapFree+"</p>" +
                "<p class=\"text-left\">交换率："+memory.swapUsage+"</p>");
        }
        // CPU信息
        var cpuInfos = json.cpuInfos;
        if(cpuInfos){
            var str = "";
            $.each(cpuInfos, function(index, value){
                str += ("<p class=\"text-left\">"+index+"用户使用率："+value.cpuUserUse+"</p>" +
                    "<p class=\"text-left\">"+index+"系统使用率："+value.cpuSysUse+"</p>" +
                    "<p class=\"text-left\">"+index+"当前等待率："+value.cpuWait+"</p>" +
                    "<p class=\"text-left\">"+index+"当前空闲率："+value.cpuFree+"</p>" +
                    "<p class=\"text-left\">"+index+"总的使用率："+value.cpuTotal+"</p>");
            });
            $("#system-info-3").html(str);
        }
        // 磁盘信息
        var diskInfos = json.diskInfos;
        if(diskInfos){
            var str = "";
            $.each(diskInfos, function(index, value){
                str += ("<p class=\"text-left\">"+index+"系统盘名称："+value.diskName+"</p>" +
                    "<p class=\"text-left\">"+index+"盘类型："+value.diskType+"</p>" +
                    "<p class=\"text-left\">"+index+"文件系统总大小："+value.diskTotal+"</p>" +
                    "<p class=\"text-left\">"+index+"文件系统剩余大小："+value.diskFree+"</p>" +
                    "<p class=\"text-left\">"+index+"文件系统已经使用量："+value.diskUse+"</p>" +
                    "<p class=\"text-left\">"+index+"内存使用率："+value.diskUsage+"</p>");
            });
            $("#system-info-4").html(str);
        }
    }
    //连接关闭的回调方法
    websocket.onclose = function(){
        websocket = null;
    }
    //监听窗口关闭事件，当窗口关闭时，主动去关闭websocket连接，防止连接还没断开就关闭窗口，server端会抛异常。
    window.onbeforeunload = function(){
        websocket.close();
    }
});

File: src/main/java/com/my/blog/website/utils/SystemInfo.java
package com.my.blog.website.utils;

import com.alibaba.fastjson.JSONObject;
import org.hyperic.sigar.*;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

/**
 * 系统信息
 *
 * @author SongM
 */
public class SystemInfo {

    public static JSONObject SystemProperty() {
        JSONObject monitorMap = new JSONObject();
        Runtime r = Runtime.getRuntime();
        Properties props = System.getProperties();
        InetAddress addr = null;
        String ip = "";
        String hostName = "";
        try {
            addr = InetAddress.getLocalHost();
        } catch (UnknownHostException e) {
            ip = "无法获取主机IP";
            hostName = "无法获取主机名";
        }
        if (null != addr) {
            try {
                ip = addr.getHostAddress();
            } catch (Exception e) {
                ip = "无法获取主机IP";
            }
            try {
                hostName = addr.getHostName();
            } catch (Exception e) {
                hostName = "无法获取主机名";
            }
        }
        // 本地ip地址
        monitorMap.put("hostIp", ip);
        // 本地主机名
        monitorMap.put("hostName", hostName);
        // 操作系统的名称
        monitorMap.put("osName", props.getProperty("os.name"));
        // 操作系统的构架
        monitorMap.put("arch", props.getProperty("os.arch"));
        // 操作系统的版本
        monitorMap.put("osVersion", props.getProperty("os.version"));
        // JVM可以使用的处理器个数
        monitorMap.put("processors", r.availableProcessors());
        // Java的运行环境版本
        monitorMap.put("javaVersion", props.getProperty("java.version"));
        // Java的运行环境供应商
        monitorMap.put("vendor", props.getProperty("java.vendor"));
        // Java供应商的URL
        monitorMap.put("javaUrl", props.getProperty("java.vendor.url"));
        // Java的安装路径
        monitorMap.put("javaHome", props.getProperty("java.home"));
        // 默认的临时文件路径
        monitorMap.put("tmpdir", props.getProperty("java.io.tmpdir"));
        return monitorMap;
    }

    private static String div(long num) {
        long gb = 1024 * 1024 * 1024;
        long mb = 1024 * 1024;
        long kb = 1024;
        if (num > gb) {
            return (BigDecimal.valueOf(num).divide(BigDecimal.valueOf(gb),2, RoundingMode.HALF_UP)) + "G";
        } else if (num > mb) {
            return (BigDecimal.valueOf(num).divide(BigDecimal.valueOf(mb),2, RoundingMode.HALF_UP)) + "M";
        } else if (num > kb) {
            return (BigDecimal.valueOf(num).divide(BigDecimal.valueOf(kb),2, RoundingMode.HALF_UP)) + "K";
        }
        return String.valueOf(num);
    }

    private static String div1(long num) {
        long gb = 1024 * 1024;
        long mb = 1024;
        if (num > gb) {
            return (BigDecimal.valueOf(num).divide(BigDecimal.valueOf(gb),2, RoundingMode.HALF_UP)) + "G";
        } else if (num > mb) {
            return (BigDecimal.valueOf(num).divide(BigDecimal.valueOf(mb),2, RoundingMode.HALF_UP)) + "M";
        }
        return String.valueOf(num);
    }

    public static JSONObject memory(Sigar sigar) {
        JSONObject monitorMap = new JSONObject();
        try {
            Runtime r = Runtime.getRuntime();
            // java总内存
            monitorMap.put("jvmTotal", div(r.totalMemory()));
            // JVM剩余内存
            monitorMap.put("jvmFree", div(r.freeMemory()));
            // JVM使用内存
            long useMemory = r.totalMemory() - r.freeMemory();
            monitorMap.put("jvmUse", div(useMemory));
            // JVM使用率
            monitorMap.put("jvmUsage", Math.round(Double.valueOf(useMemory) / Double.valueOf(r.totalMemory()) * 100) + "%");

            Mem mem = sigar.getMem();
            // 内存总量
            monitorMap.put("ramTotal", div(mem.getTotal()));
            // 当前内存使用量
            monitorMap.put("ramUse", div(mem.getUsed()));
            // 当前内存剩余量
            monitorMap.put("ramFree", div(mem.getFree()));
            // 内存使用率
            monitorMap.put("ramUsage", Math.round(Double.valueOf(mem.getUsed()) / Double.valueOf(mem.getTotal()) * 100) + "%");

            Swap swap = sigar.getSwap();
            // 交换区总量
            monitorMap.put("swapTotal", div(swap.getTotal()));
            // 当前交换区使用量
            monitorMap.put("swapUse", div(swap.getUsed()));
            // 当前交换区剩余量
            monitorMap.put("swapFree", div(swap.getFree()));
            // 交换率
            monitorMap.put("swapUsage", Math.round(Double.valueOf(swap.getUsed()) / Double.valueOf(swap.getTotal()) * 100) + "%");
        } catch (Exception e) {
        }
        return monitorMap;
    }

    public static List<JSONObject> cpuInfos(Sigar sigar) {
        List<JSONObject> monitorMaps = new ArrayList<JSONObject>();
        try {
            CpuPerc[] cpuList = sigar.getCpuPercList();
            for (CpuPerc cpuPerc : cpuList) {
                JSONObject monitorMap = new JSONObject();
                // 用户使用率
                monitorMap.put("cpuUserUse", Math.round(cpuPerc.getUser()*100) + "%");
                // 系统使用率
                monitorMap.put("cpuSysUse", Math.round(cpuPerc.getSys()*100) + "%");
                // 当前等待率
                monitorMap.put("cpuWait", Math.round(cpuPerc.getWait()*100) + "%");
                // 当前空闲率
                monitorMap.put("cpuFree", Math.round(cpuPerc.getIdle()*100) + "%");
                // 总的使用率
                monitorMap.put("cpuTotal", Math.round(cpuPerc.getCombined()*100) + "%");
                monitorMaps.add(monitorMap);
            }
        } catch (Exception e) {
        }
        return monitorMaps;
    }

    public static List<JSONObject> diskInfos(Sigar sigar) {
        List<JSONObject> monitorMaps = new ArrayList<>();
        try {
            FileSystem[] fsList = sigar.getFileSystemList();
            for (int i = 0; i < fsList.length; i++) {
                JSONObject monitorMap = new JSONObject();
                FileSystem fs = fsList[i];
                // 文件系统类型名，比如本地硬盘、光驱、网络文件系统等
                FileSystemUsage usage = sigar.getFileSystemUsage(fs.getDirName());
                switch (fs.getType()) {
                    // TYPE_UNKNOWN ：未知
                    case 0:
                        break;
                    // TYPE_NONE
                    case 1:
                        break;
                    // TYPE_LOCAL_DISK : 本地硬盘
                    case 2:
                        // 系统盘名称
                        monitorMap.put("diskName", fs.getDevName());
                        // 盘类型
                        monitorMap.put("diskType", fs.getSysTypeName());
                        // 文件系统总大小
                        monitorMap.put("diskTotal", div1(usage.getTotal()));
                        // 文件系统剩余大小
                        monitorMap.put("diskFree", div1(usage.getFree()));
                        // 文件系统已经使用量
                        monitorMap.put("diskUse", div1(usage.getUsed()));
                        // 内存使用率
                        monitorMap.put("diskUsage", Math.round(usage.getUsePercent() * 100) + "%");
                        // read
                        monitorMap.put(fs.getDevName() + "read", usage.getDiskReads());
                        // write
                        monitorMap.put(fs.getDevName() + "write", usage.getDiskWrites());
                        monitorMaps.add(monitorMap);
                        break;
                    // TYPE_NETWORK ：网络
                    case 3:
                        break;
                    // TYPE_RAM_DISK ：闪存
                    case 4:
                        break;
                    // TYPE_CDROM ：光驱
                    case 5:
                        break;
                    // TYPE_SWAP ：页面交换
                    case 6:
                        break;
                    default:
                        break;
                }
            }
        } catch (SigarException e) {
        }
        return monitorMaps;
    }

    public static void main(String[] args) {
        JSONObject serverinfo = SystemInfo.SystemProperty();
        Sigar sigar = new Sigar();
        JSONObject memory = SystemInfo.memory(sigar);
        List<JSONObject> cpuInfos = SystemInfo.cpuInfos(sigar);
        List<JSONObject> diskInfos = SystemInfo.diskInfos(sigar);
    }
}
